Raven.Client.Lightweight/Document/DocumentStore.cs

```diff
@@ -917,62 +920,62 @@ internal class NagleData
             public TaskCompletionSource<BatchResult[]> TaskCompletionSource = new TaskCompletionSource<BatchResult[]>();
         }
 
-        private readonly BlockingCollection<NagleData> nagleQueue = new BlockingCollection<NagleData>();
-        private Task nagleWriterTask;
+        private readonly ConcurrentDictionary<string, Tuple<Task, BlockingCollection<NagleData>>> nagleTasks = 
+            new ConcurrentDictionary<string, Tuple<Task, BlockingCollection<NagleData>>>();
 
-        internal Task<BatchResult[]> AddNagleData(InMemoryDocumentSessionOperations.SaveChangesData saveChangesData)
+        internal Task<BatchResult[]> AddNagleData(string databaseName, InMemoryDocumentSessionOperations.SaveChangesData saveChangesData)
         {
-            InitializeNagleQueueWriterIfNeeded();
+            var nagleQueue = GetNagleQueue(databaseName);
 
             var nagleData = new NagleData { SaveChangesData = saveChangesData };
             nagleQueue.Add(nagleData);
             return nagleData.TaskCompletionSource.Task;
         }
 
-        private void InitializeNagleQueueWriterIfNeeded()
+        private BlockingCollection<NagleData> GetNagleQueue(string databaseName)
         {
-         
-            if (nagleWriterTask != null)
-            {
-                EnsureNagleIsNotFaulted();
-                return;
-            }
-
-            lock (nagleQueue)
+            var nagle = nagleTasks.GetOrAdd(databaseName ?? "<system>", x =>
```

Comment:
This is a bad idea.
If there is contention, this can create _two_ threads for the same key.
Instead, use `Lazy<Tuple<Task, BlockingCollection<NagleData>>>`

Reply:
done